Longhand method receivers emit the wrong bound

```rust
use async_trait::async_trait;

#[async_trait]
pub trait Trait {
    async fn f(&self) {} // works
    async fn g(self: &Self) {} // does not work
}
```

```console
error[E0277]: `Self` cannot be shared between threads safely
   --> tests/test.rs:564:33
    |
564 |         async fn g(self: &Self) {}
    |                                 ^^ `Self` cannot be shared between threads safely
    |
    = help: the trait `std::marker::Sync` is not implemented for `Self`
    = note: required because of the requirements on the impl of `std::marker::Send` for `&Self`
    = note: required because it appears within the type `[static generator@tests/test.rs:564:33: 564:35 _self:&Self {}]`
    = note: required because it appears within the type `std::future::from_generator::GenFuture<[static generator@tests/test.rs:564:33: 564:35 _self:&Self {}]>`
    = note: required because it appears within the type `impl std::future::Future`
    = note: required because it appears within the type `impl std::future::Future`
    = note: required for the cast to the object type `dyn std::future::Future<Output = ()> + std::marker::Send`
```

--- a/src/expand.rs
+++ b/src/expand.rs
@@ -179,6 +179,16 @@ fn transform_sig(
                 mutability: None,
                 ..
             })) => parse_quote!(Sync),
+            Some(FnArg::Typed(arg))
+                if match (arg.pat.as_ref(), arg.ty.as_ref()) {
+                    (Pat::Ident(pat), Type::Reference(ty)) => {
+                        pat.ident == "self" && ty.mutability.is_none()
+                    }
+                    _ => false,
+                } =>
+            {
+                parse_quote!(Sync)
+            }
             _ => parse_quote!(Send),
         };
         let assume_bound = match context {

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -554,3 +554,14 @@ pub mod issue81 {
         }
     }
 }
+
+// https://github.com/dtolnay/async-trait/issues/83
+pub mod issue83 {
+    use async_trait::async_trait;
+
+    #[async_trait]
+    pub trait Trait {
+        async fn f(&self) {}
+        async fn g(self: &Self) {}
+    }
+}

